Rewrite installation and upgrade docs in README (#403)
axelboc/anki-ultimate-geography: README.md

  - [Supplementary information](#supplementary-information)
  - [Custom study](#custom-study)
- [**Getting started**](#getting-started)
- [**Upgrading**](#upgrading)
  - [First upgrade after APKG import](#first-upgrade-after-apkg-import)
  - [Keeping deck customisations](#keeping-deck-customisations)
  - [Levelling up from standard to extended](#levelling-up-from-standard-to-extended)
  - [Major version](#major-version)
- [**Deck structure**](#deck-structure)
- [**Customising the deck**](#customising-the-deck)
  - [Changes that are or can be preserved](#changes-that-are-or-can-be-preserved-)
  - [Changes that get reverted](#changes-that-get-reverted-)

>   - [Changes that prevent upgrading](#changes-that-prevent-upgrading-)

Reviewer: Useless -s at the end of the link? I guess they are no harm.

Author: I believe this is how GitHub identifies anchors to headings that end in `[space][emoji]`. I'll fix on `master` if it's not the case after merging.

1. Don't change anything in the _CrowdAnki Import Settings_ dialog box that opens -- just press _OK_ to perform the import. A dialog box should then confirm that the import was successful.

To stay informed of new releases, make sure to [watch this repository's releases](https://help.github.com/en/articles/watching-and-unwatching-releases-for-a-repository).

## Upgrading

The upgrade process is typically the same as the installation process explained in the [previous section](#getting-started). However, some situations require extra care. As a general rule, **always read the release notes carefully**; they will likely tell you what to do or point you to a page that does.

### First upgrade after APKG import

You may have initially installed _Ultimate Geography_ by importing an **APKG file**. You may have downloaded such a file from this repo or from the [deck's page](https://ankiweb.net/shared/info/2109889812) on AnkiWeb, as this used to be the recommended installation process.

If you're in this situation and wish to upgrade, proceed as follows:

1. Perform the _Getting started_ steps exactly.

> 1. You should end up with two decks: your original deck called "Ultimate Geography" and a new, duplicated deck called "Ultimate Geography_2". Every card in your original deck should have moved automatically to the new deck and your original deck should now be empty. You can verify this in Anki's [card browser](https://docs.ankiweb.net/#/browsing?id=browsing).

Reviewer 2: > your original deck should now be empty

Should we mention something like "unless you created some notes of your own"?

1. To get back to having a single deck, delete your original deck and rename the new deck to "Ultimate Geography". From here on, future upgrades will be seamless.

### Cards moved into other deck

If you've moved some of the cards out of the default _Ultimate Geography_ deck and into another deck, by default CrowdAnki will move those cards back on import.

To prevent this behaviour and update the existing cards in place, follow the steps in the [_Getting started_](@getting-started) section but in the last step, make sure to tick the _Do Not Move Existing Cards_ checkbox in the _CrowdAnki Import Settings_ dialog box.

### Standard to extended

> Importing an extended deck on top of a standard deck may bring unepected results. Instead, we recommend that you start learning the extended deck from scratch. To do so, either:

Reviewer 2: Typo: "unepected results" → "unexpected results".

If you're in this situation and wish to upgrade, proceed as follows:

1. Perform the _Getting started_ steps exactly.
1. You should end up with two decks: your original deck called "Ultimate Geography" and a new, duplicated deck called "Ultimate Geography_2". Unless you created some notes of your own, every card in your original deck should have moved automatically to the new deck and your original deck should now be empty. You can verify this in Anki's [card browser](https://docs.ankiweb.net/#/browsing?id=browsing).
1. To get back to having a single deck, move any of your own cards out of the original deck and delete it. Then, rename the new deck to "Ultimate Geography". From here on, future upgrades will be seamless.

### Keeping deck customisations

If you've made, or are thinking of making, any changes to your _Ultimate Geography_ deck, check the [_Customising the deck_](#customising-the-deck) section below to see what will happen to them the next time you upgrade, and whether you need to take any steps to preserve them.

### Levelling up from standard to extended

> Importing an extended deck on top of a standard deck may bring unexpected results. You should consider learning the extended deck from scratch. To do so, either:

Reviewer 2: If the users just click-through all the dialogues (just selecting OK), which many people will do (and I don't blame them — given the general preponderance of confirmations dialogues), then they will get unexpected results (they'll end up having a jumbled-up review schedule (with what are really new cards ("Country - Flag") presented as old cards, and vice-versa, which is arguably far worse than starting from scratch).

You're right, though, that we might wish to soften the language slightly.

Perhaps:

`may bring unexpected results` → `can be tricky and needs care`

and

`You should consider learning the extended deck from scratch` → `You may instead consider learning the extended deck from scratch`

?

- remove the standard deck first by following the instructions in the [_Major version_](#major-version) section below, or
- import the extended deck in a separate Anki profile.

If you'd rather keep your progress, proceed as follows:

1. Perform the _Getting started_ steps exactly.
1. Instead of a dialog box confirming that the import was successful, you'll see another dialog box titled _Change Note Type_. In the _Cards_ section, set _Change Flag - Country to:_ to "Flag - Country" (instead of "Country - Flag"), and set _Change Map - Country to:_ to "Map - Country" (instead of "Flag - Country").
1. Check that the other cards and fields map correctly and click on _OK_ to start the import.

### Major version

> Major versions (e.g. `v3.0`) typically indicate that upgrading may lead to a significant loss of progress, or to the loss of [some of the customisations](#customising-the-deck) you may have made to the deck that would normally be preserved. This can occur, for instance, when [significant changes](https://github.com/axelboc/anki-ultimate-geography/releases/tag/v3.0) are made to the structure of the deck, or when [lots of cards](https://github.com/axelboc/anki-ultimate-geography/releases/tag/v4.0) are removed.

Reviewer: ```suggestion
Major versions (e.g. `v3.0`) typically indicate that upgrading may lead to a significant loss of progress, or to the loss of [some of the customisations](#customising-the-deck) you may have made to the deck that would normally be preserved. This can occur, for instance, when [significant changes](https://github.com/axelboc/anki-ultimate-geography/releases/tag/v3.0) are made to the structure of the deck, or if our include criteria change which can result in [removal of some less relevant cards from the deck.](https://github.com/axelboc/anki-ultimate-geography/releases/tag/v4.0)
```

Read to me like card removal is a common thing that happens all the time! :sweat_smile:

Author: I'm actually tempted to remove this example completely. After all, we'll never remove as many cards as the last time we did. The seas criteria will only remove a couple of cards, which I don't think will justify a major release.

Author: We may remove some of the lakes with seas in their name, but again, not a huge removal.

Reviewer 2: We might also end up removing some ("geographical" rather than "political") islands.

Author: Ah right. All in all, I still don't think it will amount to "losing a significant amount of progress", especially considering that removed cards are not actually removed from people's decks on import.

I think I'll remove the quote that points to the `CONTRIBUTING` guide's _Versioning_ section. I don't think this section provides any more information than is already in the `README`.

### Standard to extended

Importing an extended deck on top of a standard deck may bring unexpected results. You should consider learning the extended deck from scratch. To do so, either:

- remove the standard deck first by following the instructions in the [_Major version_](#major-version) section below, or
- import the extended deck in a separate Anki profile.

> If you'd rather keep your progress, proceed as follows:

Reviewer 2: Looks great! (Less verbose and more clear than what I wrote!)

- Type - `UG::Sovereign_State`, `UG::Oceans+Seas`, `UG::Continents`
- Continent - `UG::Africa`, `UG::Asia`, `UG::Europe`, `UG::North_America`, `UG::Oceania`, `UG::South_America`
- Region - `UG::Caribbean`, `UG::Middle_East`, `UG::Southeast_Asia`, `UG::European_Union`, `UG::Mediterranean`, `UG::East_Africa`, `UG::West_Africa`

## Customising the deck

Interested in making changes to the deck in Anki? Here's what you need to know.

### Changes that are or can be preserved 💖

- **Adding your own cards and media files**

>   Go wild. 🌊 Upgrading will not affect any new cards you create, or any new media files you add.

Reviewer: User added media files will not be deleted from their Anki by ugrading, however their inclusion in any of the existing cards in the fields themselves can and probably will be overwritten.

Reviewer: And then if they are not references Anki will remove them on a Media check.

Reviewer 2: Perhaps:

```suggestion
  Go wild. 🌊 Upgrading will not affect any new cards you create, or any new media files you add to the newly added cards.
```

?

  If a duplicated deck is [created on import](#first-upgrade-after-apkg-import), your own cards will stay in their original deck.

- **Moving cards into another deck**

  Some people prefer to combine multiple decks into one to review them together.

>   By default, CrowdAnki will move all the cards back into the _Ultimate Geography_ deck on import. To prevent this behaviour and update the existing cards in place, follow the steps in the [_Getting started_](@getting-started) section but in the last step, make sure to tick the _Do Not Move Existing Cards_ checkbox in the _CrowdAnki Import Settings_ dialog box.

Reviewer: ```suggestion
  By default, CrowdAnki will move all the cards back into the _Ultimate Geography_ deck on import. To prevent this behaviour make sure to tick the _Do Not Move Existing Cards_ checkbox in the _CrowdAnki Import Settings_ dialog box to update the existing cards in place, when following the steps in the [_Getting started_](@getting-started) section.
```

Give the pertinent info asap. Helps greatly in these long ReadMe files when people are scanning for the info the want.

Author: Good call, though I think "and update the existing cards in place" needs to stay put for the sentence to remain clear:

> To prevent this behaviour and update the existing cards in place, make sure to tick the _Do Not Move Existing Cards_ checkbox in the _CrowdAnki Import Settings_ dialog box, when following the steps in the [_Getting started_](@getting-started) section.

  > Note that if you had previously deleted the _Ultimate Geography_ deck, it will be recreated and may contain new cards.

- **Changing the deck's options**

  Deck options are, for instance, the number of new cards per day, the maximum number of reviews per day, etc.

  When importing the deck for the first time, CrowdAnki creates a new "option group" called _Ultimate Geography_ (unless it already exists). If you've made any changes to this group's options, they will be reverted on import.

  To avoid losing your options, create your own option group and give it a distinctive name. After you upgrade, you'll just have to switch the deck back to this option group.

- **Customising the templates and styles**

  Any edits you make to the built-in templates of the _Ultimate Geography_ note type will be reverted on upgrade. Same goes with the CSS styles that are shared by all the templates.

>   One way to workaround this and preserve your changes is to  first duplicate the note type and switch all the cards to the new note type in the Anki browser. You can then make your template and style changes in the duplicate note type without affecting the original note type. The next time you upgrade, all the cards will switch back to the original note type, but you can easily switch them back to your customised note type in the Anki browser.

Reviewer 2: Yes, this is far better (and more future-proof) than my approach!

Perhaps we should phrase it as "cloning" the note model, as that's how it's called in the interface as well as in the manual? Also, should we link the manual in case people don't know how to do it (I barely remembered and I've done it before...)?

```suggestion
  One way to workaround this and preserve your changes is to first [clone the note type](https://docs.ankiweb.net/#/editing?id=adding-a-note-type) and switch all the cards to the new note type in the Anki browser. You can then make your template and style changes in the cloned note type without affecting the original note type. The next time you upgrade, all the cards will switch back to the original note type, but you can easily switch them back to your customised note type in the Anki browser.
```

(It's possible that I'm missing an easier way of duplicating note models...)

<hr/>

I've tested it, and it works as desired and expected.

However, I'm slightly worried, as the `crowdanki_uuid` of the "cloned" note model is the same as that of the original one. (I've checked by exporting the deck with the cloned note model and by inspecting `collection.anki2` with sqlite.) (This is analogous to https://github.com/Stvad/CrowdAnki/issues/118, just for note models...) Hence, in principle, `CrowdAnki` could "update" the new (cloned) note model, rather than the original.

Fortunately, it seems (looking at [the UUID code](https://github.com/Stvad/CrowdAnki/blob/a583fa8ceaa4932cf00630476e357ef2e420233e/crowd_anki/utils/uuid.py#L34)) that `CrowdAnki` finds the first matching note_model (the first one having a matching UUID), and assuming that Anki keeps its note_models in chronological order, this will be the original note_model (as we desire), so there _shouldn't_ be problems (but I'm still not 100% comfortable...).

Author: Hmm, that's disappointing. That being said, the worst that can happen is that CrowdAnki updates the cloned note type, in which case the user would just lose their customisation. Maybe I can just add a warning that this may not work in the future or something and to recommend backing up the deck before upgrading.

Reviewer 2: > this may not work in the future

Hopefully, I'll get around to fixing https://github.com/Stvad/CrowdAnki/issues/118 before sufficient things change in the Anki internals that this stops working.

> recommend backing up the deck before upgrading

FWIW `CrowdAnki` does a backup before every import, but doing one manually won't hurt.

  > Note that you may have to update your custom templates and styles afterwards to stay in line with the built-in templates.

- **Adding or removing a template**

  We're getting into dangerous territory here, but you should be able to make this work by duplicating the default note type as explained in the previous point, _Customising the templates and their styles_.

  The difference is that, when you import the deck and/or when you switch back the cards to your customised note type, you may see a dialog box asking you to match the templates of one note type to the templates of the other note type. (This same dialog box appears when importing the extended deck [on top of the standard deck](#levelling-up-from-standard-to-extended).)

>   As long as you map the templates that Anki asks you to match correctly, you should be fine. However, we strongly recommend that you try this out in a separate Anki profile first.

Reviewer 2: tldr; Do not recommend :scream:

I might be doing something wrong, but I don't see how upgrading a deck with an added card template can be done without loss of progress (without manually fiddling with the `deck.json` beforehand (my workaround of first exporting your current deck and then re-importing doesn't really work (or perhaps it works, while crashing Anki??)!)).

With the UUID cloning issue, the note model is (for me) updated "in-place" without any dialogue, unlike in the Standard → Extended case (or unlike if the original model had had an extra card added), since here 

1. the UUIDs of the two note models are the same
2. the original note model is not changed (no cards/fields added/removed)

Hence, all the notes are silently moved back to the "original" note model and silently have one card template removed, (without the corresponding "extra"/now attached cards removed). As a result, when I open the browser, I get a "list index out of range" error (since there are cards attached to the "original" note model which have an index higher than the number of cards/note...) and can't do anything. If I try to import an export of the deck with the cloned note model, then I get a key error (again because there are cards that have no right to exist...) (though after killing Anki and restarting, it seems that all cards including the ones from the added templates are restored, again, with progress). Alternatively, (instead of trying to import an export of the deck with the cloned note) I can check the database, but that deletes the extra cards.

If the UUID cloning issue were not present, (i.e. the cloned deck had its own UUID, rather copying that of the original) then you'd be presented with the standard note model migration dialogue (like for Standard → Extended) and would have to force mapping the added cards (those corresponding to the added template) to "Nothing"...

<hr/>

Removing templates should be fine.

Author: You're right, I get the error as well. I think when I tested, I must have added one template and removed another, which led to the same number of templates and may explain why I didn't see the error.

Author: Removing templates by cloning the note type ends up being more complicated than just removing the template again after the upgrade. So I'll move it into the next section.

- **Renaming the deck or changing its description**

  Upgrading always restores the deck's default name and description.

- **Editing the content of a note**

  Examples:

  - adding a link to Wikipedia in the _Country info_ field of Curaçao;
  - adding your own memoization hint to the _Flag similarity_ field of Poland;
  - adding or removing a tag from a note;
  - renaming a tag.

  > If you correct a typo or a mistake, don't forget to [open an issue](https://github.com/axelboc/anki-ultimate-geography/issues/new) or a Pull Request so you don't lose your correction the next time you upgrade. 😉

- **Deleting a card**

  Upgrading will bring the card back. If you don't want to learn a card at all or if you want to delay learning it, consider:

  - suspending or burying the card;
  - creating a [filtered deck](#custom-study) that doesn't include the card;
  - moving the card into an "icebox" deck and ticking the _Do Not Move Existing Cards_ the next time you upgrade, as explained in the previous section, under _Moving cards into another deck_.

### Changes that prevent upgrading ⛔

> - **Adding a new field (population, currency, etc.)**

Reviewer 2: Ugh, yeah, you're right. This (the fact that you get a `list index out of range` error rather than having the extra field successfully mapped to Nothing) is a `CrowdAnki` bug. :(

>   Adding a new field to the _Ultimate Geography_ note type will prevent you from upgrading the deck entirely. When you'll try to import it again, Anki will likely error and you'll risk losing your entire progress. Hopefully Anki, CrowdAnki, Brain Brew and UG will [one day](https://github.com/ohare93/brain-brew/issues/4#issuecomment-644975261) find a way to make this possible.

Reviewer: ```suggestion
  Adding a new field to the _Ultimate Geography_ note type will prevent you from upgrading the deck entirely. When you'll try to import it again, Anki will likely error and you'll risk losing your entire progress. Hopefully Anki, CrowdAnki, Brain Brew, and UG will [one day](https://github.com/ohare93/brain-brew/issues/4#issuecomment-644975261) find a way to make this possible.
```

Oxford Comma! :wink: 

And god damn, linking people to the mega thread is just mean :rofl: :+1:
